### PR TITLE
Endpoints to get a list of assets which overlap a point, box, or quadkey

### PR DESCRIPTION
**Pull request by mackdelany**
_Background for this PR can be found here: https://github.com/developmentseed/titiler/discussions/350_

This PR adds three new endpoints to Titiler:

```
http://<titiler-host>/mosaicjson/{minx},{miny},{maxx},{maxy}/assets
http://<titiler-host>/mosaicjson/{lon},{lat}/assets
http://<titiler-host>/mosaicjson/{quadkey}/assets
```

Each endpoint returns a list of cogs which overlap the geometries given, e.g:

```
[
      "s3://super-awesome-cog-1.tif",
      "s3://super-awesome-cog-2.tif",
      "s3://super-awesome-cog-3.tif"
]
```

**Re: mercantile**
I believe this PR adds [mercantile](https://github.com/mapbox/mercantile) as a dependency, I am happy to change this if desired! I'm used to mercantile + saw it used in some of the notebooks (so I assumed it might be ok to use). If we'd like to remove mercantile as a dependency I can do one of the following:
- Add methods to return mercator tiles for the given geometries
- Use other functions that maintainers / @vincentsarago / @geospatial-jeff recommend

And of course - happy to make any other changes / edits desired

### CODE REVIEW

File: src/titiler/mosaic/titiler/mosaic/factory.py [hunk 1]
```diff
@@ -484,3 +486,65 @@ def validate(self):
...
+                    for x in range(tl_tile.x, br_tile.x + 1)
+                    for y in range(tl_tile.y, br_tile.y + 1)
+                ]
+                assets = list(
+                    {
+                        asset
+                        for asset_list in [mosaic.assets_for_tile(*t) for t in tiles]
+                        for asset in asset_list
+                    }
+                )
```
Comment (vincentsarago) [inline src/titiler/mosaic/titiler/mosaic/factory.py:519]: ❤️

**Comment by vincentsarago** [comment]
thanks @mackdelany everything looks good to me 🙏 

> I believe this PR adds mercantile as a dependency, I am happy to change this if desired! I'm used to mercantile + saw it used in some of the notebooks (so I assumed it might be ok to use).

**mercantile** is already a dependency for cogeo-mosaic so this is totally fine. The major blocker for us to remove it, is to add quadkey method to morecantile (https://github.com/developmentseed/morecantile/issues/19).

**Comment by vincentsarago** [comment]
@mackdelany Before we merge this PR here is what we need:
- update changelog 
- update docs https://github.com/developmentseed/titiler/blob/master/docs/endpoints/mosaic.md
- add mercantile to `titiler.mosaic` setup.requirement (just a safer approach in case we remove mercantile in cogeo-mosaic at one point)

**Comment by mackdelany** [comment]
Great, thanks @vincentsarago 👌 🥳

I've updated the changelog, docs and `titiler.mosaic` requirement.

Let me if any changes are needed / anything else you need from me!

File: CHANGES.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 # Release Notes
 
+## 0.3.5 (2021-08-05)
```
Comment (vincentsarago) [inline CHANGES.md:3]:
```suggestion
## 0.3.5 (TBD)
```

File: CHANGES.md [hunk 3]
```diff
@@ -1,5 +1,11 @@
 # Release Notes
 
+## 0.3.5 (TBD)
+
+### titiler.mosaic
+
+* add `/{quadkey}/assets`, `/{lon},{lat}/assets`, `/{minx},{miny},{maxx},{maxy}/assets` GET endpoints to return a list of assets that intersect a given geometry (https://github.com/developmentseed/titiler/pull/351)
```
Comment (vincentsarago) [inline CHANGES.md:7]:
```suggestion
* add `/{quadkey}/assets`, `/{lon},{lat}/assets`, `/{minx},{miny},{maxx},{maxy}/assets` GET endpoints to return a list of assets that intersect a given geometry (author @mackdelany, https://github.com/developmentseed/titiler/pull/351)
```

File: src/titiler/mosaic/tests/test_factory.py [hunk 4]
```diff
@@ -142,3 +142,32 @@ def test_MosaicTilerFactory():
...
         )
         assert response.status_code == 200
+
+        response = client.get("/mosaic/0302302/assets", params={"url": mosaic_file},)
+        assert response.status_code == 200
+        assert all(
+            filepath.split("/")[-1] in ["cog1.tif"] for filepath in response.json()
+        )
+
+        response = client.get("/mosaic/-71,46/assets", params={"url": mosaic_file},)
```
Comment (vincentsarago) [inline src/titiler/mosaic/tests/test_factory.py:152]:
```suggestion
        response = client.get("/mosaic/-71,46/assets", params={"url": mosaic_file})
```

**Comment by vincentsarago** [comment]
🙏 thanks @mackdelany, everything is perfect

**Pull request merged**